Fix GC uid races and handling of conflicting ownerReferences

liggitt opened the pull request:
Addresses issues with race conditions encountered when observing invalid ownerReferences

Fixes #65200

- [x] address review comments
- [x] add comments to each commit on the scenario being addresses
- [x] add unit tests for GC controller and graph builder fixes
- [x] ensure the coverage of the following testcases are represented:

----

- [x] child in namespace A with owner reference to namespaced type in namespace B
    * should be deleted immediately
    * event should be logged in namespace A with involvedObject of bad-child indicating the error
    * test both orders:
      * real parent added to graph, then bad child added to graph
      * bad child added to graph, virtual parent added to graph, then real parent added to graph

- [x] child that is cluster-scoped with owner reference to namespaced type in namespace B
    * should not be deleted
    * event should be logged in namespace kube-system with involvedObject of bad-child indicating the error
    * test both orders:
      * real parent added to graph, then bad child added to graph
      * bad child added to graph, virtual parent added to graph, then real parent added to graph

- [x] child pointing at non-preferred still-served apiVersion of parent object (e.g. rbac/v1beta1)
    * should not be deleted prematurely
    * should not repeatedly poll attemptToDelete while waiting
    * should be deleted when the actual parent is deleted

- [x] child pointing at no-longer-served apiVersion of still-existing parent object (e.g. extensions/v1beta1 deployment)
    * should not be deleted (this is indistinguishable from referencing an unknown kind/version)
    * should not repeatedly poll attemptToDelete once real parent is observed
    * TODO: follow up: should be fixed up by a controller adjusting known ownerReference API versions

- [x] child pointing at no-longer-served apiVersion of no-longer-existing parent object (e.g. extensions/v1beta1 deployment)
    * should not be deleted (this is indistinguishable from referencing an unknown kind/version)
    * should repeatedly poll attemptToDelete
    * should not block deletion of legitimate children of missing deployment
    * TODO: follow up: should be fixed up by a controller adjusting known ownerReference API versions and once fixed up, should be deleted since parent does not exist

- [x] child pointing at incorrect apiVersion/kind of still-existing parent object (e.g. core/v1 Secret with uid=123, where an apps/v1 Deployment with  uid=123 exists)
    * should be deleted immediately
    * should not trigger deletion of legitimate children of parent

---

Follow-ups:
- [x] ~~add a controller/mechanism to migrate apiVersions of ownerReferences for known kubernetes types that have graduated or moved API groups (this is a pre-existing problem)~~ - tracked in https://github.com/kubernetes/kubernetes/issues/96650
- [x] before 1.20 release - write a tool that scans items in the cluster and identifies invalid ownerReferences that would be impacted by this change, update the release note to describe running that tool on a 1.19 cluster to determine if there is existing bad data - https://github.com/kubernetes-sigs/kubectl-check-ownerreferences

/kind bug
/cc @jpbetz @deads2k 
/sig api-machinery

```release-note
Resolves non-deterministic behavior of the garbage collection controller when ownerReferences with incorrect data are encountered. Events with a reason of `OwnerRefInvalidNamespace` are recorded when namespace mismatches between child and owner objects are detected. The [kubectl-check-ownerreferences](https://github.com/kubernetes-sigs/kubectl-check-ownerreferences) tool can be run prior to upgrading to locate existing objects with invalid ownerReferences.
* A namespaced object with an ownerReference referencing a uid of a namespaced kind which does not exist in the same namespace is now consistently treated as though that owner does not exist, and the child object is deleted.
* A cluster-scoped object with an ownerReference referencing a uid of a namespaced kind is now consistently treated as though that owner is not resolvable, and the child object is ignored by the garbage collector.
```

liggitt committed:
Add GC integration race test

liggitt committed:
Switch GC absentOwnerCache to full reference

Before deleting an object based on absent owners, GC verifies absence of those owners with a live lookup.

The coordinates used to perform that live lookup are the ones specified in the ownerReference of the child.

In order to performantly delete multiple children from the same parent (e.g. 1000 pods from a replicaset),
a 404 response to a lookup is cached in absentOwnerCache.

Previously, the cache was a simple uid set. However, since children can disagree on the coordinates
that should be used to look up a given uid, the cache should record the exact coordinates verified absent.
This is a [apiVersion, kind, namespace, name, uid] tuple.

liggitt committed:
Avoid marking virtual nodes as observed when they haven't been

Virtual nodes can be added to the GC graph in order to represent objects
which have not been observed via an informer, but are referenced via ownerReferences.

These virtual nodes are requeued into attemptToDelete until they are observed via an informer,
or successfully verified absent via a live lookup. Previously, both of those code paths
called markObserved() to stop requeuing into attemptToDelete.

Because it is useful to know whether a particular node has been observed via
a real informer event, this commit does the following:

* adds a `virtual bool` attribute to graph events so we know which ones came from a real informer
* limits the markObserved() call to the code path where a real informer event is observed
* uses an alternative mechanism to stop requeueing into attemptToDelete when a virtual node is verified absent via a live lookup

liggitt committed:
Replace virtual node with observed node if identity differs

If the graph contains a virtual node (because some child object referenced it in an OwnerRef),
and a real informer event is observed for that uid at different coordinates,
we want to fix the coordinates of the node in the graph to match the actual coordinates.

The safe way to do this is to clone the node, replace the identity in the clone,
then replace the node with the clone.

Modifying the identity directly is not safe because it is accessed lock-free from many code paths.

Replacing the node in the graph from processGraphChanges is safe because it is the only graph writer.

liggitt committed:
Enqueue dependents for deletion when their ownerReference does not match observed parent coordinates

When adding a dependent to the graph, we ensure there is a node representing each owner reference,
and add the dependent to each parent node.

If the parent node already exists, and the dependent's ownerReference
coordinates disagree with the verified coordinates, add the dependent to the attemptToDelete queue.

This queue will check the dependent's ownerReferences using the coordinates specified by the dependent.
If all of the owners can be verified absent, the dependent will be deleted.
If some are still present, or if there are errors looking them up, the dependent will not be deleted.

If the parent node has been observed via informer event (so we know the coordinates are accurate),
and the verified owner is namespaced, and the dependent is not in the same namespace,
an event will be recorded for user visibility, since cross-namespace ownerReferences are not supported.

liggitt commented:
/retest

liggitt commented:
/skip

fedebongio commented:
/assign @jpbetz 
/cc @caesarxuchao

liggitt committed:
Refactor identityFromEvent

liggitt committed:
Handle virtual delete events when children don't agree on owner coordinates

If a virtual delete event is received for a node whose dependents disagree on the parent's coordinates:
1. propagate the delete to children that matched the verified absent coordinates
2. if the existing node is virtual, select a new set of coordinates from the remaining dependents
3. do not delete the parent node from the graph if the parent node is non-virtual,
   or if there are dependents that do not agree with the virtual delete event coordinates

liggitt committed:
Make node removal conditional in processGraphChanges

liggitt commented:
/retest

mattfarina commented:
/retest pull-kubernetes-conformance-kind-ga-only-parallel

mattfarina commented:
/test pull-kubernetes-conformance-kind-ga-only-parallel

liggitt commented:
review for this is scheduled for 10/8

liggitt committed:
Short-circuit attemptToDelete loop for virtual nodes that are removed or observed

Virtual nodes are added to the attemptToDelete queue, and continue getting requeued
until they are successfully verified absent or are observed via informer.

In the meantime, if the real object associated with that UID is observed via informer,
or is observed to be deleted via informer, the graph node for that UID can be removed
or marked as observed. In that case, we should stop retrying to get the virtual node coordinates.

liggitt commented:
/retest

liggitt reviewed (commented):
comments from review with @deads2k

liggitt committed:
Add GC unit tests

Adds unit tests covering the problematic scenarios identified
around conflicting data in child owner references

                      Before   After
package level         51%      68%
garbagecollector.go   60%      75%
graph_builder.go      50%      81%
graph.go              50%      68%

Added/improved coverage of key functions that had lacking unit test coverage:

* attemptToDeleteWorker
* attemptToDeleteItem
* processGraphChanges (added coverage of all added code)

dims commented:
cc @nikhita

mkorbi commented:
Hello! This PR has been updated frequently in the last weeks, but it looks like it stuck so I'd like to check what's the status. The code freeze is starting 12.11.2020 (about 3-4 weeks from now) and while there is still plenty of time, we want to ensure that each PR has a chance to be merged on time.
As the PR is tagged for 1.20, is it still planned for this release?

liggitt commented:
Yes, it is still planned. I'm working through the remaining items currently.

liggitt committed:
Queue non-matching children for deletion when a virtual node is marked as observed

When we observe valid coordinates for a previously virtual node,
if there are dependents that do not agree with those coordinates,
add them to the attemptToDelete queue.

This queue will check the dependent's ownerReferences using the coordinates specified by the dependent.
If all of the owners can be verified absent, the dependent will be deleted.
If some are still present, or if there are errors looking them up, the dependent will not be deleted.

If the verified owner is namespaced, and the dependent is not in the same namespace,
an event will be recorded for user visibility, since cross-namespace ownerReferences are not supported.

mkorbi commented:
Hey @liggitt short reminder, code freeze will be by today

soltysh commented:
> Hey @liggitt short reminder, code freeze will be by today

You mean Thursday 11/12

liggitt commented:
> > Hey @liggitt short reminder, code freeze will be by today
> 
> You mean Thursday 11/12

definitely thursday, three more development days left (including today)

liggitt commented on pkg/controller/garbagecollector/graph_builder.go:
```diff
@@ -565,6 +565,25 @@ func (gb *GraphBuilder) processGraphChanges() bool {
+		if observedIdentity != existingNode.identity {
+			// make a copy (so we don't modify the existing node in place), store the observed identity, and replace the virtual node
+			klog.V(5).Infof(
+				"GraphBuilder replace virtual node %s %s/%s, namespace %s, name %s with observed node %s/%s, namespace %s, name %s",
```
raised to v2, logged event

liggitt commented on pkg/controller/garbagecollector/graph_builder.go:
```diff
@@ -366,6 +367,21 @@ func (gb *GraphBuilder) addDependentToOwners(n *node, owners []metav1.OwnerRefer
+				// n's owner reference doesn't match the observed identity.
+				// add n to the attemptToDelete queue since it has a potentially invalid owner reference.
+				addedToAttemptToDelete = true
+				klog.V(5).Infof("node %s references an owner %s with coordinates that do not match the observed identity", n.identity, ownerNode.identity)
```
raised to v2 here and below, logged an event below for a positive namespace mismatch

liggitt commented on pkg/controller/garbagecollector/graph_builder.go:
```diff
@@ -366,6 +367,21 @@ func (gb *GraphBuilder) addDependentToOwners(n *node, owners []metav1.OwnerRefer
+				// n's namespace doesn't match the observed owner namespace.
+				// add n to the delete queue since it has an invalid owner reference.
+				addedToAttemptToDelete = true
+				klog.V(5).Infof("node %s references an owner %s but does not match namespaces", n.identity, ownerNode.identity)
```
v2 + event

liggitt committed:
Plumb event recorder to garbage collector controller

liggitt committed:
Log cluster-scoped owners referencing namespaced owners, avoid retrying lookups forever

If a cluster-scoped dependent references a namespace-scoped owner,
this is an invalid relationship, and the lookup will never succeed in attemptToDelete.

Short-circuit requeueing in attemptToDelete and log.

liggitt commented on pkg/controller/garbagecollector/graph_builder.go:
```diff
@@ -603,6 +602,7 @@ func (gb *GraphBuilder) processGraphChanges() bool {
 			existingNode.identity = observedIdentity
 			gb.uidToNode.Write(existingNode)
 		}
+		existingNode.markObserved()
```
moved markObserved after cloning/replacing, so we don't mark the virtual node that disagreed with our coordinates as observed

liggitt commented on pkg/controller/garbagecollector/graph_builder.go:
```diff
@@ -372,23 +374,56 @@ func (gb *GraphBuilder) addDependentToOwners(n *node, owners []metav1.OwnerRefer
 			gb.attemptToDelete.Add(ownerNode)
 		} else if !addedToAttemptToDelete && ownerNode.isObserved() {
 			// The owner node has been observed via an informer
```
the diff is unpleasant because this changed the order to check for namespace mismatch first (since we record an event for that scenario and can surface more useful info to the user)

droslean commented:
@liggitt Do we still consider this to land on 1.20?

liggitt commented:
> @liggitt Do we still consider this to land on 1.20?

It's on the bubble. It's a fix for a rare but severe bug.

liggitt commented:
/assign @deads2k

liggitt commented:
Comments addressed and tests added. Left additional commits as fixups for easier review, will squash fixup commmits once review is complete.

deads2k commented:
/lgtm

/hold

please squash into some reasonable number of commits if possible.  If not possible, I'll still take it.

jpbetz reviewed (commented):
Have 2 very minor comments and one question.

This PR is in excellent shape. The comments on the commits are a huge help.

liggitt commented on pkg/controller/garbagecollector/graph_builder.go:
```diff
@@ -366,6 +367,21 @@ func (gb *GraphBuilder) addDependentToOwners(n *node, owners []metav1.OwnerRefer
+				addedToAttemptToDelete = true
+				klog.V(5).Infof("node %s references an owner %s with coordinates that do not match the observed identity", n.identity, ownerNode.identity)
+				gb.attemptToDelete.Add(n)
+			} else if len(ownerNode.identity.Namespace) > 0 && ownerNode.identity.Namespace != n.identity.Namespace {
```
captured in `ownerIsNamespaced` var and added comment:

```
// the dependent's namespace doesn't match the observed owner's namespace, this is definitely wrong.
// cluster-scoped owners can be referenced as an owner from any namespace or cluster-scoped object.
```

liggitt commented on pkg/controller/garbagecollector/graph_builder.go:
```diff
@@ -366,6 +367,21 @@ func (gb *GraphBuilder) addDependentToOwners(n *node, owners []metav1.OwnerRefer
+			// The owner node has been observed via an informer
+			if !ownerReferenceMatchesCoordinates(owner, ownerNode.identity.OwnerReference) {
+				// n's owner reference doesn't match the observed identity.
+				// add n to the attemptToDelete queue since it has a potentially invalid owner reference.
```
clarified:
```
// Enqueue the potentially invalid dependent node into attemptToDelete.
// The garbage processor will verify whether the owner references are dangling
// and delete the dependent if all owner references are confirmed absent.
```

liggitt commented on pkg/controller/garbagecollector/graph_builder.go:
```diff
@@ -565,6 +565,25 @@ func (gb *GraphBuilder) processGraphChanges() bool {
+		if observedIdentity != existingNode.identity {
+			// make a copy (so we don't modify the existing node in place), store the observed identity, and replace the virtual node
+			klog.V(5).Infof(
+				"GraphBuilder replace virtual node %s %s/%s, namespace %s, name %s with observed node %s/%s, namespace %s, name %s",
```
markObserved() on line 657 sets virtual=false

jpbetz replied:
Oh, I think `markObserved()` should be responsible for setting virtual. I also found https://github.com/kubernetes/kubernetes/pull/92743/files#r522420492 which explains how that logic has been corrected.

jpbetz commented:
/lgtm

Thanks for driving this to completion @liggitt! The test coverage is outstanding.

liggitt commented:
squashed down to commits where each commit documents and solves one problem.

zero diff from what was reviewed by @deads2k and @jpbetz: https://github.com/kubernetes/kubernetes/compare/32331d1831c53ce1a05b1f0280120e9c6fb35e66..e491c3bc7056530d82590d95f0af0e8c4d8dded5

deads2k commented:
/lgtm
/hold cancel

It's time.

liggitt commented:
docs update for 1.20 at https://github.com/kubernetes/website/pull/25091

krmayankk commented:
curious why did we make the cross namespace reference an invalid case ?

liggitt commented:
> why did we make the cross namespace reference an invalid case ?

namespaces are intended to be independent of each other, so cross-namespace references have not been permitted in things like ownerReferences, secret/configmap volume references, etc.

krmayankk commented:
> > why did we make the cross namespace reference an invalid case ?
> 
> namespaces are intended to be independent of each other, so cross-namespace references have not been permitted in things like ownerReferences, secret/configmap volume references, etc.
> 
> additionally, granting permissions to namespace `a` is not generally intended to provide visibility or ability to interact with objects from namespace `b` (or cause system controllers to interact with objects from namespace `b`).

thanks

little-guy-lxr commented:
could this fix merge to 1.19.x ?

liggitt commented:
> could this fix merge to 1.19.x ?

No, this is a significant enough rewrite of the garbagecontroller that it is limited to 1.20+

For clusters on previous versions, I would recommend running https://github.com/kubernetes-sigs/kubectl-check-ownerreferences tool is available to identify invalid ownerReferences that could trigger the bug this PR is fixing so they can be eliminated.
